content/systems/filecoin_blockchain/struct/tipset/_index.md

```diff
@@ -26,4 +27,20 @@ Due to network propagation delay, it is possible for a miner in epoch N+1 to omi
 
 Block producers are expected to coordinate how they select messages for inclusion in blocks in order to avoid duplicates and thus maximize their expected earnings from transaction fees (see [Message Pool](message_pool)).
 
-{{<embed src="../chain/tipset.id" lang="go" >}}
\ No newline at end of file
+The main Tipset structure in the Lotus implementation includes the following:
+
+{{<embed src="/externals/lotus/chain/types/tipset.go"  lang="go" symbol="TipSet">}}
+
+
+Semantic validation of a Tipset is performed in:
+
+```go
+func NewTipSet(blks []*BlockHeader) (*TipSet, error)
+```
+
+Semantic validation includes the following checks:
+- The tipset should be composed of at least one block
+- All blocks within a tipset should have the same height
+- All blocks within a tipset have the same parents, both in terms of number of parents and in terms of CIDs.
+
+The Lotus implementation of the `Tipset` can be found [here](https://github.com/filecoin-project/lotus/blob/master/chain/types/tipset.go).
```

Comment:
no need to include this the `TipSet` code block above already has a perma link to github